### PR TITLE
initial cadc-datalink-server library

### PR DESCRIPTION
**Pull request by pdowler**

### CODE REVIEW

File: cadc-datalink-server/src/main/java/org/opencadc/datalink/DataLink.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    public enum Term { // TODO: re-use the VocabularyTerm code once extracted from caom2
+        THIS("#this"),
+        AUXILIARY("#auxiliary"),
+        PREVIEW("#preview"),
+        PROGENITOR("#progenitor"),
+        DERIVATION("#derivation"),
+        PROC("#proc"),
+        CUTOUT("#cutout"),
+        THUMBNAIL("http://www.openadc.org/caom2#thumbnail"), // TODO: move these out
+        PKG("http://www.openadc.org/caom2#pkg");             // TODO: move these out
```
Comment (at88mph) [inline cadc-datalink-server/src/main/java/org/opencadc/datalink/DataLink.java]: There is a typo in the hostname.
`openadc` should be `opencadc`.

Comment (pdowler) [reply cadc-datalink-server/src/main/java/org/opencadc/datalink/DataLink.java]: fixed

File: cadc-datalink-server/src/main/java/org/opencadc/datalink/ServiceParameter.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    public void setValueRef(String val, String ref) {
+        if (val != null && ref != null) {
+            throw new IllegalArgumentException("only one of value and ref may be set");
+        }
+        this.value = val;
+        this.ref = ref;
+    }
+
+    public String getValue() {
+        if (value == null) {
```
[the remark this answers is not in the document]
Comment (pdowler) [reply cadc-datalink-server/src/main/java/org/opencadc/datalink/ServiceParameter.java]: VOTable param has to have a value, so null has to be serialised as zero-length string; ID/IDREF attrs can only be added if they are non-null so the check is elsewhere. Probably this null -> "" should be elsewhere.

File: cadc-datalink-server/src/main/java/org/opencadc/datalink/server/DataLinkSource.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+     * links or more loosely related resources.
+     *
+     * @param dlo true to generate download links only
+     */
+    void setDownloadOnly(boolean dlo);
+
+    /**
+     * Set maximum number of links to return. If maxrec is specified, truncate the
+     * returned link iterator when the specified number of links is reached plus
+     * any remaining links from the same input ID such that if an ID is returned
```
[the remark this answers is not in the document]
Comment (pdowler) [reply cadc-datalink-server/src/main/java/org/opencadc/datalink/server/DataLinkSource.java]: One input ID gives 1+ output links; just dn't want to truncate in ten middle of handling an ID and give some of it's  links.

**Pull request merged**